File: src/services/shared-java-lib/src/main/java/com/shareconnectsave/shared/ErrorResponse.java
package com.shareconnectsave.shared;

// Pattern: Polyglot contract — this record and the C# record at
// services/user-service/Contracts/ErrorResponse.cs serialize to the identical JSON
// shape defined once in contracts/openapi/error-envelope.yaml. Neither language's
// type is the source of truth for the wire format; the YAML is, and both records are
// derived from it independently. A Java record's component names already serialize as
// plain lower-camelCase field names via Jackson with zero configuration, which is why
// this type needs no annotations to match the contract — unlike the C# side, whose
// serializer defaults to PascalCase and therefore needs [JsonPropertyName] to match.
public record ErrorResponse(String code, String message, String traceId) {
}

File: src/services/user-service/Contracts/ErrorResponse.cs
using System.Text.Json.Serialization;

namespace user_service.Contracts;

// Pattern: Polyglot contract, enforced explicitly rather than by convention.
// Reason: System.Text.Json's default naming policy is PascalCase (Code, Message,
// TraceId). The Java side of this contract (shared-java-lib's ErrorResponse record)
// serializes plain lower-camelCase field names with no naming policy involved at all.
// If this record relied on ambient JsonSerializerOptions to produce the right casing,
// a future change to Program.cs's serializer config (in this service or copied into
// another one) would silently break the wire shape for every client. [JsonPropertyName]
// pins the JSON keys to contracts/openapi/error-envelope.yaml regardless of whatever
// serializer settings this or any other .NET service configures.
//
// This is the canonical shape — copied (not referenced) into each other .NET service
// in their own tickets, per T003's scope.
public record ErrorResponse(
    [property: JsonPropertyName("code")] string Code,
    [property: JsonPropertyName("message")] string Message,
    [property: JsonPropertyName("traceId")] string TraceId
);
